### PR TITLE
Add ToastMessageComponent with tests

### PR DESCRIPTION
**Pull request by lsparlin**
`ToastMessageComponent` uses the "Slots" feature of View Components to offer flexible display options

## Screenshots
image

image

image

### CODE REVIEW

**Comment by thubamamba** [comment]
LGTM. 👍🏽

Maybe we can also think of use cases where a form has multiple form fields.

**Comment by lsparlin** [comment]
Sadly ViewComponent has some known issues with rails form_with https://viewcomponent.org/known_issues.html

**Comment by jacobdaddario** [comment]
> Sadly ViewComponent has some known issues with rails form_with https://viewcomponent.org/known_issues.html

Yeah, I recently looked into this for a personal project. Since Rails' form builder is essentially a component, it creates some problems. The design of `FormBuilder` being a wrapper around `*_tag` methods creates a lot of weird code interactions in general 😵

File: app/components/toast_message_component.html.erb
Comment (Tonksthebear) [inline app/components/toast_message_component.html.erb:4]: Just a heads up, the tailwind purger won't correctly purge classes that are spliced together like this. It's annoying, but you'd have to clearly declare text-red-400, text-green-400, etc. for the purger to pick up on it, rather than concatenating strings together

Comment (Tonksthebear) [reply app/components/toast_message_component.html.erb:4]: I noticed there was at least one other spot where the tailwind class was concatenated like this, so just thought I'd mention it!

Comment (lsparlin) [reply app/components/toast_message_component.html.erb:4]: Oh thanks for pointing this out. I'm not super familiar with tailwind yet. Do you have any suggestions?

Comment (joemasilotti) [reply app/components/toast_message_component.html.erb:4]: Great catch, @Tonksthebear.

@lsparlin, instead of passing in only the `color`, we should pass in the entire class name. For example, `text-red-400` or `text-blue-400`. Then we can make sure that the Tailwind config is checking `app/components/` for class names to purge.

Comment (Tonksthebear) [reply app/components/toast_message_component.html.erb:4]: I'm not sure if it's proper, but I normally just define a list of constant class colors in the ruby file, such as:

`STYLE = {
  blue: 'text-blue-400',
  green: 'text-green-400'
}`

Then I just pass in the style I want to the component like you already have set up. Just instead of setting @color the parameter, I'd do
`@color = STYLES[color]`

Then swap `text-#{@color}-400` with `@color`

Hopefully that makes sense! I wrote this up pretty quickly... there might be a better way, just wanted to get some ideas out there in case they're useful.

Comment (lsparlin) [reply app/components/toast_message_component.html.erb:4]: There are several such classes in the component. You think there should be a parameter for each of them?

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This looks great outside of the Tailwind classes comment from @Tonksthebear. Once that is resolved we can merge this in.

File: app/components/toast_message_component.html.erb
Comment (joemasilotti) [reply app/components/toast_message_component.html.erb:4]: Yeah, this could get ugly quickly, @lsparlin. I wonder what other apps are doing in this space.

Comment (Tonksthebear) [reply app/components/toast_message_component.html.erb:4]: I definitely understand how it becomes unwieldy. I just couldn't find another way because Tailwind requires such a verbose syntax. To accommodate multiple parameters, I just make the hash larger and define getters. So some quick sudo code for what I do

```
STYLES = {
  blue: {
    header: 'text-blue-200',
    message: 'text-blue-400',
    footer: 'text-blue-100'
  }
...
def header_classes
  STYLES[@color][:header]
end

def message_classes
  STYLES[@color][:message]
end

def footer_classes
  STYLES[@color][:message]
end
```

Then I would just call the corresponding getters from the html.

Again, I completely get how verbose this solution may sound but this is what worked best for my mental model when working with Tailwind

Comment (lsparlin) [reply app/components/toast_message_component.html.erb:4]: So as long as the class (full literal string) shows up anywhere in `html` or `rb` files tailwind will be happy?

Thank you @Tonksthebear

Comment (Tonksthebear) [reply app/components/toast_message_component.html.erb:4]: Not to overload this thread with more information, but another avenue I've pursued is creating a helper that pulls from a config file for the corresponding classes.

For instance, I'd have a config/styles.yml file that contains something like

```
toast:
  blue:
    header: 'text-blue-200 ...'
    message: 'text-blue-400 ...'
    footer: 'text-blue-100 ...'
```

Then in the html I'd just call `tailwind(toast: { @color => :header })`

I've enjoyed this pattern due to no boilerplate in the .rb files, but it does feel dangerously close to just simulating css classes. I suppose another avenue would be using ViewComponent's sidecar assets in some way (if possible). 

In either case, it's a problem I've been personally wrestling with so I'd love to hear other solutions

Comment (Tonksthebear) [reply app/components/toast_message_component.html.erb:4]: > So as long as the class (full literal string) shows up anywhere in `html` or `rb` files tailwind will be happy?
> 
> Thank you @Tonksthebear

Yup! Just have to update the tailwind.config.js to point it to the correct files

Comment (joemasilotti) [reply app/components/toast_message_component.html.erb:4]: My vote is for a big hash and helper methods for now. But I do love a YAML-powered configuration for the different types of colors! Maybe if we need them in the future that will be a good route to take.

**Comment by lsparlin** [comment]
Finally got around to this. Hope I wasn't holding anything up. This should work better with Tailwind's purging

**Comment by joemasilotti** [comment]
Amazing work, team! This code is looking awesome. I pulled in `main` and made two style changes.

Before I merge, can you fix the display issue of the component? It somehow lost its bottom margin, or something.

image

**Comment by lsparlin** [comment]
@joemasilotti Weird, it still has the `mb-8` class

**Comment by joemasilotti** [comment]
I know... I'm stumped too. But I don't think this was an issue before the latest round of commits.

**Comment by lsparlin** [comment]
Got it. Needed to make tailwind are of `components/**/*.html.erb`

**Comment by joemasilotti** [comment]
Amazing! Will merge when CI is green. Thanks again for this!

**Pull request merged**
